### PR TITLE
featureCounts upgraded to 1.5.3

### PR DESCRIPTION
**Pull request by AsierGonzalez**

### CODE REVIEW

**Comment by AjitPS** [comment]
Thanks @AsierGonzalez

**Comment by bgruening** [comment]
@AsierGonzalez it seems you also need to update the test data.

**Comment by AsierGonzalez** [comment]
Hi @bgruening,

I'm afraid I do not understand why I should update the test data. Do you mean I should run or add some test?

Thanks

**Comment by nsoranzo** [comment]
@AsierGonzalez featureCounts changed slightly its output format, see the failed Travis job log: https://travis-ci.org/galaxyproject/tools-iuc/jobs/287071336#L1696

To update the test data you can run `planemo test` with the `--update_test_data` option, then commit and push the changes in the `tools/featurecounts/test-data/` .

Finally, can you remove the line:
```
-S  $pe_parameters.orientation
```
from the command and the `orientation` param from the inputs, since it has been deprecated in the new version?

**Comment by AsierGonzalez** [comment]
Hi @nsoranzo,

I have been working with @AjitPS on this and we have been unable to make it work. We've run `planemo test --galaxy_root=~/galaxy --update_test_data featurecounts.xml` and we get the following error:

`AttributeError: 'module' object has no attribute 'python_2_unicode_compatible'
Testing complete. HTML report is in "~/shed_tools/toolshed.g2.bx.psu.edu/repos/iuc/featurecounts/de2c5e5a206c/featurecounts/tool_test_output.html".
cp: cannot stat ‘/tmp/tmp_BibeD/jobfiles/*’: No such file or directory
Cannot locate xUnit report option for tests - update Galaxy for more detailed breakdown.
One or more tests failed. See ~/shed_tools/toolshed.g2.bx.psu.edu/repos/iuc/featurecounts/de2c5e5a206c/featurecounts/tool_test_output.html for detailed breakdown.`

In addition, tool_test_output.html is empty, apparently because `Uncaught TypeError: Cannot read property 'num_tests' of undefined`.

Do you have any recommendations? I don't know know what has changed in the output between the previous version and this one and whether Ithis may mean that something needs to be changed in the wrapper.

**Comment by AsierGonzalez** [comment]
I see, I was too naive when I tought that a one-line change would suffice...

I am completely new to wrapper development and testing used in galaxy. I will look at the material available and see if I can make it work.

**Comment by nsoranzo** [comment]
@AsierGonzalez I've pushed a commit to your fork to update the test data, now the Travis build passes. `planemo test --update_test_data /path/to/tools/featurecounts/` worked fine for me as usual, so I suspect something is broken in your Planemo setup.

@yhoogstrate Would it be OK to merge this and then add the new parameters in follow-up PRs?

**Comment by AjitPS** [comment]
@AsierGonzalez could you test it again now on our planemo, with the new changes by Nicola ?

**Comment by AsierGonzalez** [comment]
Thanks for the changes @nsoranzo. We need to investigate further why planemo failed, what I do not understand is whether we first need to run featureCounts to create the test ouput data or if the `--update_test_data` is supposed to deal with that?

Anyway, I am currently looking at the options that they have added to featureCounts since 1.4.6-p5 and that may be worth including. For example, there is `-J` to count exon-exon junction reads that some people may want to use. And `-L` for counting long reads. I was thinking to include this two, are there any rules or caveats to include parameters?

**Comment by yhoogstrate** [comment]
It is OK to patch since the patch won't do any harm. But I would prefer someone creates an issue that some arguments are missing. Did I mis something, for example did somebody run into a bug?

**Comment by yhoogstrate** [comment]
@AsierGonzalez I was thinking of the same options. Usually IUC tries to offer all options which at least make sense to be used in Galaxy. For example sometimes tools generate files in inconvenient formats which are not included in Galaxy and settings related to that have no reason to be included. The `-J` an `-L` can be really useful I think. Thanks for all the work !

**Comment by AsierGonzalez** [comment]
Yes, last week I startted getting "silent" errors that were not caught by galaxy (perhaps because they were not reported in stderr) that disappeared in the latest version. In addition, there was a notable performance improvement, with jobs that used to run for around 1h finishing in 8 mins.

**Comment by AsierGonzalez** [comment]
Thank @yhoogstrate, I will add these parameters then.

**Comment by nsoranzo** [comment]
> what I do not understand is whether we first need to run featureCounts to create the test output data or if the `--update_test_data` is supposed to deal with that?

`planemo test` runs the Galaxy tests, which run featureCounts. The results of each test are compared with the expected results (which are stored in the `test-data/` subdirectory) and the test will fail if they are different.
When you add the `--update_test_data` parameter, Planemo will still fail the test, but the files in `test-data/` are updated with the actual results of the Galaxy job run by Planemo. So running again `planemo test` (without `--update_test_data`) should result in all tests passing.
Doing a `git diff` you can see the modifications made by Planemo in the `test-data` subdir, and if they make sense, commit and push them.

@yhoogstrate I just didn't want to add extra work for a first-time contributor, but it seems @AsierGonzalez is willing to add the 2 new options, so all good!

File: tools/featurecounts/featurecounts.xml [hunk 1]
```diff
@@ -28,6 +28,12 @@
                     $extended_parameters.multimapping_enabled.fraction
                 #end if
 
+                $extended_parameters.exon_exon_junction_read_counting_enabled.count_exon_exon_junction_reads
+                #if str($extended_parameters.exon_exon_junction_read_counting_enabled.count_exon_exon_junction_reads) == " -J"
+                    -G "$extended_parameters.exon_exon_junction_read_counting_enabled.genome"
```
[the remark this answers is not in the document]
Comment (nsoranzo) [reply tools/featurecounts/featurecounts.xml]:
```
#end if
```
is missing here.

File: tools/featurecounts/featurecounts.xml [hunk 2]
```diff
@@ -80,6 +86,10 @@
             && cut -f 1,6 body.txt > "${output_feature_lengths}"
         #end if
 
+        #if str($extended_parameters.exon_exon_junction_read_counting_enabled.count_exon_exon_junction_reads) == " -J"
+            && tail -n+2 "output.jcounts" > "${output_jcounts}"
```
Comment (nsoranzo) [inline tools/featurecounts/featurecounts.xml]: Single quotes.

File: tools/featurecounts/featurecounts.xml [hunk 3]
```diff
@@ -367,6 +405,15 @@
                 <action name="column_names" type="metadata" default="Feature,Length" />
             </actions>
         </data>
+
+        <data format="tabular"
+              name="output_jcounts"
+              label="${tool.name} on ${on_string}: junction counts">
+              <filter>extended_parameters.exon_exon_junction_read_counting_enabled.count_exon_exon_junction_reads == " -J"</filter>
```
Comment (nsoranzo) [inline tools/featurecounts/featurecounts.xml]: Indentation of this line is wrong.

File: tools/featurecounts/featurecounts.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nsoranzo) [reply tools/featurecounts/featurecounts.xml]: Planemo doesn't check the Cheetah syntax (i.e. the content of `<command>`, `<version_command>` and `<configfile>` elements).

File: tools/featurecounts/featurecounts.xml [hunk 4]
```diff
@@ -258,6 +268,34 @@
                    label="Minimum mapping quality per read"
                    help="The minimum mapping quality score a read must satisfy in order to be counted. For paired-end reads, at least one end should satisfy this criteria. 12 by default." />
 
+            <conditional name="exon_exon_junction_read_counting_enabled">
+                <param name="count_exon_exon_junction_reads"
+                       type="boolean"
+                       truevalue=" -J"
```
[the remark this answers is not in the document]
Comment (nsoranzo) [reply tools/featurecounts/featurecounts.xml]: No, it won't, Galaxy substitutes newlines in `<command>` with a whitespace.

**Comment by AsierGonzalez** [comment]
This is odd. I have made the changes that Nicola suggested but, from what I see in the travis report, one of the three tests failed.

**Comment by bgruening** [comment]
All "good", this is not your error. We try to fix it.

**Comment by nsoranzo** [comment]
Thanks a lot @AsierGonzalez!

**Comment by AsierGonzalez** [comment]
Thanks to you @nsoranzo for all the support. I will do some testing now to make sure that it works as expected.

**Comment by AsierGonzalez** [comment]
I think I have done something wrong with the filtering of the junction counting option (line 400):

```
<filter>extended_parameters.exon_exon_junction_read_counting_enabled.count_exon_exon_junction_reads == " -J"</filter>
```
I get an (obviously empty) junction counts output in history even if I do not choose the option. I think the sintax is wrong but what I don't understand is why this comparison returns `true` (as it is included in the output).

Anyway, according to the [filter documentation](https://docs.galaxyproject.org/en/master/dev/schema.html#tool-outputs-data-filter)  "Variables that are part of conditionals are accessed using a hash named after the conditional" so I wonder if that line should look more something like:

```
<filter>exon_exon_junction_read_counting_enabled['count_exon_exon_junction_reads'] == "-J"</filter>
```

In addition, this parameters is located within the `extended_parameters` section, so I don't know whether this should also be factored in the filter.

Thanks

**Comment by nsoranzo** [comment]
First issue is that there's still a space in `" -J"` (I think you already spotted that).

Yes, `extended_parameters` should be also used, I think the correct syntax should be:

```xml
<filter>extended_parameters['exon_exon_junction_read_counting_enabled']['count_exon_exon_junction_reads'] == '-J'</filter>
```

It would be good to add `expect_num_outputs` to the tests, see https://planemo.readthedocs.io/en/latest/writing_how_do_i.html#test-output-filters-work

Can you open a PR with these fixes (don't need to increase the tool version)?

**Comment by AsierGonzalez** [comment]
Thanks Nicola! As you said I saw the extra space in `" -J"` and tested it without it but it didn't work. I will try your suggestions and open a new PR.

**Pull request merged**